Moved things related to positions from "placed_tile.go" to "position.go"

Lenetis opened the pull request.

bstrzelecki commented:
Since position in now in a separate file, we can separate this into its own package, because its most commonly used feature anyway.

BognaLew reviewed (changes requested):
Please add tests for Add and PositionFromSide.

bstrzelecki commented on pkg/game/elements/position.go:
```diff
@@ -0,0 +1,65 @@
+/*
+Returns relative position directed by the side.
+Caution! It is supposed to be used with side directing only one cardinal direction (or two edges connected by corner)!
+Otherwise it will return undesired value!
```
We might also want to handle the case when the function would return undesired value.

Lenetis replied:
TBH, I'm not a fan of this idea. It adds a lot of complexity to a very simple function.
Maybe that's just a premature optimisation thought, but I have a feeling iterating through all primary sides to check if more than one is set everytime we call this function would make the engine run slower.

bstrzelecki replied:
Well, I'm not a fan of methods that require more lines of documentation designated for how this function can fail than how it's actually used. This method incorrectly handles invalid cases and minimal solution would be to just test for strict equallity (i.e. side == sidemod.Top) so that the default case would be actually useful where one could panic or return identity. Test cases on how invalid cases are handled would also be welcomed.

bstrzelecki replied:
Well, I'm not a fan of methods that require more lines of documentation designated for how this function can fail than how it's actually used. This method incorrectly handles invalid cases and minimal solution would be to just test for strict equallity (i.e. side == sidemod.Top) so that the default case would be actually useful where one could panic or return identity. Test cases on how invalid cases are handled would also be welcomed.

Lenetis replied:
Added panic on incorrect input, and updated the documentation comment to better reflect what the function actually does

Lenetis commented:
- Moved position to its own package
- Added tests
- Simplified the position.FromSide() function - I used an older version of this function and I'm not sure why it was changed. Please take a look at commit [3969cb3](https://github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pull/42/commits/3969cb396f4b80af437acf86716180dcedb86ea5) because maybe I'm overlooking some edge case here.

bstrzelecki commented on pkg/game/position/position.go:
```diff
@@ -0,0 +1,70 @@
+position.FromSide(side.Right) ---> (1,0)
+position.FromSide(side.BottomLeftEdge) == position.FromSide(side.BottomRightEdge) ---> (0, -1)
+*/
+func FromSide(side sideMod.Side) Position {
+	primarySides := 0
+	for _, checkedSide := range []sideMod.Side{sideMod.Top, sideMod.Right, sideMod.Left, sideMod.Bottom} {
+		if side&checkedSide != 0 {
+			primarySides++
+		}
+	}
+
+	if primarySides == 0 {
+		return New(0, 0)
+	} else if primarySides == 1 {
+		switch {
+		case side&sideMod.Top != 0:
+			return New(0, 1)
+		case side&sideMod.Right != 0:
+			return New(1, 0)
+		case side&sideMod.Left != 0:
+			return New(-1, 0)
+		case side&sideMod.Bottom != 0:
+			return New(0, -1)
+		}
+	}
+	panic(fmt.Sprintf("position.FromSide called with more than one primary side. 'side' = %08b", side))
+}
```
```suggestion
func FromSide(side sideMod.Side) Position {
	switch side {
	case sideMod.Top:
		return New(0, 1)
	case sideMod.Right:
		return New(1, 0)
	case sideMod.Left:
		return New(-1, 0)
	case sideMod.Bottom:
		return New(0, -1)
	default:
		panic(fmt.Sprintf("position.FromSide called with more than one primary side. 'side' = %08b", side))
	}
}
```

Lenetis replied:
This won't work as intended, as for example (0, 1) should be returned not only when side=Top, but also when it's equal to TopRightCorner and TopLeftCorner. That's why there is & and checking if it's != 0

(see the documentation comment for more details)

bstrzelecki replied:
Ok, then only tests that feature invalid cases (e.g. combination of sideMod.Left and sideMod.BottomRightEdge) are missing.

Lenetis replied:
They are not missing.
In the `TestPositionFromSidePanic` function (in position_test.go) the following invalid cases are tested:
```
side.Top | side.Left,
side.RightBottomEdge | side.BottomRightEdge,
side.Top | side.Right | side.Bottom | side.Left,
side.Top | side.BottomLeftEdge,
```
